### PR TITLE
Replication logic

### PR DESCRIPTION
**Pull request by pinzart90**
Original PR - which was reverted due to failures - https://github.com/DynamoDS/Dynamo/pull/13125#issue-1309521124

This PR implements replication logic (part of it anyway)
Most of the replication functions are copied from the old VM. The new functions are mostly located right under the original ones.

New structures introduced:

CLRStackValue => equivalent to StackValue
Wraps the CLR objects so that we can store useful metadata (like rank, ProtoCore.Type info etc)
CLRFunctionEndpoint => equivalent to FunctionEndpoint
Wraps the MethodBase objects and stores useful metadata (like rank, ProtoCore.Type info etc)
MSILRuntimeCore - equivalent to RuntimeCore
Holds type information like ClassTable. Makes it easier to mimic the old VM functions
Marshaler changes:
I introduced new virtual methods (that can handle CLRStackValues) to the existing Marshaler classes ...so there should be no breaking changes.

The deferred work is marked in code with TODO_MSIL. I will create jira tasks once we finalize the PR
TODO : Add task to support int32 data types.

### CODE REVIEW

File: src/Engine/EmitMSIL/CodeGenIL.cs
Comment (mjkkirschner) [inline src/Engine/EmitMSIL/CodeGenIL.cs:1432]: this is because `stepNode` always has a value?

File: src/Engine/EmitMSIL/CodeGenIL.cs
Comment (mjkkirschner) [inline src/Engine/EmitMSIL/CodeGenIL.cs:1441]: EmitFunctionCall could do something special for builtin nodes - as it does when it encounters operators or indexing calls...

File: src/Engine/EmitMSIL/Replication.cs
Comment (mjkkirschner) [inline src/Engine/EmitMSIL/Replication.cs:62]: if we can avoid writing our own GC, I think we no longer need this flag.

File: src/Engine/ProtoCore/Lang/FunctionEndPoint.cs
Comment (mjkkirschner) [inline src/Engine/ProtoCore/Lang/FunctionEndPoint.cs:159]: does this comment still apply?

File: src/Engine/ProtoCore/Lang/Replication/Replicator.cs
Comment (mjkkirschner) [inline src/Engine/ProtoCore/Lang/Replication/Replicator.cs:343]: spelling (lose)

File: src/Engine/ProtoCore/Lang/Type.cs
Comment (mjkkirschner) [inline src/Engine/ProtoCore/Lang/Type.cs:139]: maybe will cause issues with tests

File: test/CodeGenILTests/ReplicationTests.cs
Comment (mjkkirschner) [inline test/CodeGenILTests/ReplicationTests.cs:14]: you should be able to inherit from microTest and you'll get the setup / teardown for free.

File: src/Engine/EmitMSIL/CodeGenIL.cs
Comment (pinzart90) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1441]: I will add a task for this case

File: src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs
Comment (pinzart90) [inline src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs:1935]: Always return hash `0` for null values ..

Comment (pinzart90) [reply src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs:1935]: I hope that does not clash with another hash code...

File: src/Engine/ProtoCore/Lang/FunctionEndPoint.cs
Comment (pinzart90) [reply src/Engine/ProtoCore/Lang/FunctionEndPoint.cs:159]: I did not try to adapt the old functions to the new VM except for the bare minimum.
We can adjust comments and redundancies later on

File: src/Engine/ProtoCore/Lang/Replication/Replicator.cs
Comment (pinzart90) [reply src/Engine/ProtoCore/Lang/Replication/Replicator.cs:414]: copy pasted form the old VM (I only tried to adapt the bare minimum)
We can do spell checks later (There are definitely not the only ones)

File: src/Engine/ProtoCore/Lang/Type.cs
Comment (pinzart90) [reply src/Engine/ProtoCore/Lang/Type.cs]: removed

File: test/CodeGenILTests/ReplicationTests.cs
Comment (pinzart90) [reply test/CodeGenILTests/ReplicationTests.cs]: done

**Comment by pinzart90** [comment]
@mjkkirschner @aparajit-pratap FYI
Since the revert:
I implemented a couple more replication functions (bare minimum) so that the existing tests would pass

File: src/Engine/EmitMSIL/CodeGenIL.cs
Comment (mjkkirschner) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1179]: hmm, thats okay with me, I hope we can implement compile time function calls without replication soon though.

**Review by mjkkirschner — APPROVED** [review]
Hey @pinzart90 - some thoughts on this:

1. In general, I think this is a reasonable approach to get to a state where we can test performance and correctness of many more graphs.
2. I think this introduces a ton of complexity that I was excited to leave behind in the new design - but maybe that was wishful thinking.
3. I think it would be good as part of this work to immediately bring in some .dyn tests into the performance folder so we can take advantage of this work to get metrics for more complex graphs vs the old vm.
4. I am curious for your thoughts on where to go next with replication specifically. For example, now that we have similar runtime replication functionality as in the old VM - does it make sense to try shrinking these new classes, tearing out anything we can to while maintaining correctness? Or does it make more sense to leave this runtime behavior as is, and focus instead on putting as much cleverness as possible into the compiler (compile time replication, compile time direct function calls, specific type info from function calls etc)

**Comment by pinzart90** [comment]
> Hey @pinzart90 - some thoughts on this:
> 
> 1. In general, I think this is a reasonable approach to get to a state where we can test performance and correctness of many more graphs.
> 2. I think this introduces a ton of complexity that I was excited to leave behind in the new design - but maybe that was wishful thinking.
> 3. I think it would be good as part of this work to immediately bring in some .dyn tests into the performance folder so we can take advantage of this work (and @sm6srw's) to get metrics for more complex graphs vs the old vm.
> 4. I am curious for your thoughts on where to go next with replication specifically. For example, now that we have similar runtime replication functionality as in the old VM - does it make sense to try shrinking these new classes, tearing out anything we can while maintaining correctness? Or does it make more sense to leave this runtime behavior as is, and focus instead on putting as much cleverness as possible into the compiler (compile time replication, compile time direct function calls, specific type info from function calls etc)?

My intention (for the short term - until we get performance metrics) was to get replication working as much as possible like the old VM, without too much investment in making it leaner/simpler.
If the new VM does inherently provide simpler logic (like GC), then we can definitely  use that from the start (instead of bringing in the more complicated parts of the old VM). It did not seem to me like the new VM brought anything to the table regarding replication. If I missed something in this category, then let me know and I can take another look.

My thoughts on next steps are:
a. Get the new VM (replication) to as state where it can do as much as possible so that we can compare performance.
b. Do not bring in redundant logic from the old VM (like GC). Use the new VM inherent capabilities as much as possible 
c. Once performance is recorded and we get approval to proceed, then start to think about ways to make replication simpler/leaner and easier to understand (with a close eye on what features we want to keep working)

File: test/CodeGenILTests/ReplicationTests.cs
Comment (aparajit-pratap) [inline test/CodeGenILTests/ReplicationTests.cs:220]: I think the result should be:
0,2,4
0,2,5

0,3,4
0,3,5

1,2,4
1,2,5

1,3,4
1,3,5

File: src/Engine/EmitMSIL/CodeGenIL.cs
Comment (aparajit-pratap) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1179]: In that case, can we remove the implementation for `ValueAtIndexDynamic` form `BuiltIn.cs`?

Comment (aparajit-pratap) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1179]: @pinzart90, won't we have the same issue with the builtin operators that @sm6srw added in this case, where we are replicating over those operators?

File: test/CodeGenILTests/ReplicationTests.cs
Comment (pinzart90) [reply test/CodeGenILTests/ReplicationTests.cs:220]: yes...that looks correct.
I'll take a look

Comment (pinzart90) [reply test/CodeGenILTests/ReplicationTests.cs:220]: the test assertion function was wrong. Fixed directly in branch

File: src/Engine/EmitMSIL/CodeGenIL.cs
Comment (pinzart90) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1179]: yes we should remove `ValueAtIndexDynamic`.
If @sm6srw 's builtin operators are only supposed to handle rank 0 arguments...then all should be fine. If you pass it a list of args, then it will (and should) replicate.

I did not see any arrays or lists to be handled in those builtin operators

Comment (aparajit-pratap) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1179]: Ok, so as long as the parameter is supposed to be a single value and is of type object, it's fine, got it.

File: src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs
Comment (pinzart90) [reply src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs:684]: Just a refactor so that we can share more code with the CLRStackValue

File: src/Engine/ProtoCore/Lang/FunctionEndPoint.cs
Comment (pinzart90) [reply src/Engine/ProtoCore/Lang/FunctionEndPoint.cs:334]: we can

**Pull request merged**
